### DISCUSSION TITLE
Reliable way to get size of WT's viewport?

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#13980] [General]
In the simple scenario ... my app run by a Windows shell in OpenConsole.exe in WT on the local machine ... is there a reliable way to get the number of rows and the number of columns in WT's viewport?  Thanks!

**Comment by DHowett** [comment]
`GetConsoleScreenBufferInfo` will continue to return the size of the viewport in `srWindow`, just as it does in a console host session.

**Comment by vefatica** [comment]
MODE.EXE will screw that up.  If you use  "mode con: lines=large_number" srWindow changes to the largest number of lines that will fit on the monitor, though nothing has visually changed in WT.

**Reply by DHowett** [reply]
_please try to reply in the appropriate sub-thread using the `Write a reply` box_

It'll screw that up until we reject it or handle it by resizing the Terminal window, at which point `GetConsoleScreenBufferInfo` will resume working as normal. That's tracked in #5094.

**Reply by vefatica** [reply]
`please try to reply in the appropriate sub-thread using the Write a reply box`

I know better and forget to do it (as I almost did just now with your reminder staring me in the face!).  :-)

**Reply by DHowett** [reply]
Discussions are strange, for sure.

So, the console screen buffer thing is definitely just... we haven't figured out the best thing to do for `MODE` in a multi-tenant window (like, we don't want one pane to be able to cause a resize of other adjacent panes!), so right now we just lie to applications and tell them that it "succeeded" and try to muddle through.

This causes trouble, because things like `wmic.exe` call the moral equivalent of `mode cols=9999` on startup. Terminal _hates_ this.

**Reply by vefatica** [reply]
As you probably know, other things get screwed up too.  For example, absolute cursor positioning (ESC[r;cH) uses the new srWindow. allowing rows that are out of bounds in WT.  And TCC, my usual shell, gets hopelessly messed up.

Moral of the story: don't do that!

**Reply by j4james** [reply]
> we don't want one pane to be able to cause a resize of other adjacent panes!

I've been considering this recently, and I'm starting to think the best approach might be the second option in #5094 (i.e. accept the requested size as a kind of virtual window, similar to the way conhost works when the buffer width and window width are different). This lets the app keep working as expected, and the user can always resize the pane/window afterwards to get rid of any scrollbars.

This would also make the #1860 fans happy.

**Reply by vefatica** [reply]
I've never been a fan of a horizintal scroll bar.  And even though I don't follow issues closely, I somehow got the impression that a horizontal scroll bar was definitely not going to happen in WT.

**Reply by zadjii-msft** [reply]
If it was a "definitely not going to happen", I'd close it out. I had thoughts in https://github.com/microsoft/terminal/issues/1860#issuecomment-738814484 that I've used to keep my mind open for the last two years. 

I dunno, I'm still kinda of the impression we should reject client resizes if there are more than one pane[1]. The "virtual width" with a horizontal scrollbar - that kinda can make sense, sure. But then what about increasing the height? How could we indicate that?

[1]: Though, like, what if you zoomed a single pane, then ran `mode` - should that resize the Terminal? I guess maybe.

**Reply by j4james** [reply]
> The "virtual width" with a horizontal scrollbar - that kinda can make sense, sure. But then what about increasing the height? How could we indicate that?

Yeah, a virtual height is a bit awkward, but the way I've seen other terminals handle this is with something like a dotted line in the scrollback marking the separation between the top of the page and the rest of the scrollback. I know it's not a great UI, but it's got to be better than the page layout breaking completely.

Of course if users don't like that they can always disable the option, and then we just ignore size requests altogether. But that only really works on apps that are prepared to handle their size request being ignored.

Not that I'm in any hurry to implement this, but it's one of those things I'd like to get to eventually, because it's key part of the VT windowing extension (which includes the paging operations we were discussing recently).
